lib/zzhc/zzhc.c

```diff
@@ -0,0 +1,600 @@
+#define AT_RESPONSE_LEN   64        /** Length of AT-command response */
+#define HTTP_PKT_LEN      512       /** Length of HTTP request packet */
+#define RX_BUF_LEN        128       /** Length of Rx Buffer */
+#define JSON_MAX_LEN      320       /** Length of json */
+#define IMEI_LEN          15        /** Length of IMEI */
+#define ICCID_LEN         20        /** Length of ICCID */
+#define RETRY_TIMEOUT     (3600)      /** 1 hour, 3600s */
+#define RETRY_CNT_MAX     3         /** Max. value of retry counter */
+
+/**@brief State machine */
+enum state {
+	STATE_INIT = 0,
+	STATE_WAIT_FOR_REG,
+	STATE_START,
+	STATE_HTTP_REQ_SENT,
+	STATE_HTTP_RESP_RECVD,
+	STATE_RETRY_TIMEOUT,
+	STATE_END,
+};
+
+/**@brief Context type of self-registration (zzhc) */
+struct zzhc {
+	enum state state;               /**< State machine. */
+
+	bool registered;                /**< Whether registered to operator */
+	bool sim_ready;                 /**< Whether simcard is ready */
+	int  sock_fd;                   /**< Socket file descriptor */
+	int  http_pkt_len;              /**< HTTP packet length */
+	int  rx_offset;                 /**< RX offset */
+	int  delimiter;                 /**< Delimiter buffer */
+	bool header_recvd;              /**< Header received */
+	int  retry_cnt;                 /**< Retry counter */
+	char ueiccid[ICCID_LEN+1];      /**< Locally stored ICCID */
+	char at_resp[AT_RESPONSE_LEN];  /**< Rx buffer for AT-cmd response */
+	char http_pkt[HTTP_PKT_LEN];    /**< Buffer to compile HTTP packets */
+
+	void *thread;                   /**< Pointer to thread context */
+	void *sem;                      /**< Pointer to Semophore */
+	void *at_list;                  /**< List of AT-parameters */
+	void *storage;                  /**< Pointer to storage */
+};
+
+/**@brief Host name of self-registration server. */
+static char const hostname[] = "zzhc.vnet.cn";
+
+/**@brief +CEREG event prefix. */
+static const char status_cereg[7] = "+CEREG:";
+
+/**@brief Lookup table for chip revision. */
+static const char chip_rev[4] = "XA01";
+
+#include "zzhc_port.c"
+
+/**@brief Function to handle AT-command notifications. */
+static void at_notif_handler(void *context, char *response)
+{
+	int rc;
+	struct zzhc *ctx = (struct zzhc *)context;
+
+	if (ctx == NULL) {
+		return;
+	}
+
+	if (memcmp(response, status_cereg, sizeof(status_cereg)) == 0) {
+		rc = get_cereg_evt(ctx, response);
+		if (rc < 0) {
+			return;
+		}
+
+		LOG_DBG("+CEREG: %d", rc);
+		if (rc == 1 || rc == 5) {
+			LOG_DBG("Registered to operator network");
+			ctx->registered = true;
+			sem_post(ctx->sem);
+		} else {
+			LOG_DBG("De-registered from operator network");
+			ctx->registered = false;
+		}
+	}
+}
+
+/**@brief Search and replace. */
+static void search_char_and_replace(char *data, char tok, char substitute)
+{
+	do {
+		char *cptr = strchr(data, tok);
+
+		if (cptr == NULL) {
+			break;
+		}
+		*cptr = substitute;
+	} while (true);
+}
```

Comment:
This is O(N*N).

Let's do a single pass ( O(N) ).

for(int i = 0; i < strlen(data); i++)
{ if (data[i] == tok) data[i] = substitue; }

Reply:
I prefer calling library as much as possible:
`	char *cptr = data;

	while (cptr != NULL) {
		if ((cptr = strchr(cptr, tok)) != NULL) {
			*cptr++ = substitute;
		}
	}`

Reply:
I find the version without strchr easier to read (and probably faster assuming strchr is not inlined), but at least it's now O(N), so if you prefer using strchr then that is OK.

Reply:
The solution having strlen() is 2-pass. data is traversed two times: the first time searching for NULL, and the second time searching for tok.

(Disclaimer: I'm not expert in algorithms...... )

Reply:
My mistake, I thought strlen was O(1).

using the revised strchr version is fine.